### ISSUE TITLE
Incorrect boolean mask in flash attention

### ISSUE THREAD

**Issue by stoprightthere** [lucidrains/x-transformers#160]
Hi,

Looks like [flash_attn](https://github.com/lucidrains/x-transformers/blob/4fd11a530b9b15ff8528fe5a34ae02c3d03d4817/x_transformers/attend.py#L105) requires `mask` to be True whenever a position should NOT be attended to. However, `F.scaled_dot_product_attention` expects the inverse (True means that the position takes part in the attention), if the mask is boolean.

Unless there is `attn_bias`, there is no conversion, so essentially the mask is incorrect.

tl;dr: flip `mask` whenever it exists and `attn_bias` does not (and after applying the causal mask).

**Comment by lucidrains** [comment]
@stoprightthere shoot :man_facepalming: 

could you see if https://github.com/lucidrains/x-transformers/commit/2a5bbdd83b74b61c944093b5d238b70aba196303 looks better?

**Comment by lucidrains** [comment]
@stoprightthere yea, i am keeping with the convention of `True` for attend, `False` for mask out

**Comment by stoprightthere** [comment]
Thanks for the quick reply!

I think it should then be flipped in [here](https://github.com/lucidrains/x-transformers/blob/2a5bbdd83b74b61c944093b5d238b70aba196303/x_transformers/x_transformers.py#L830C31-L830C31) and then [here](https://github.com/lucidrains/x-transformers/blob/4fd11a530b9b15ff8528fe5a34ae02c3d03d4817/x_transformers/attend.py#L227), possibly also other places :) Because you flip the mask input to `Attention` [here](https://github.com/lucidrains/x-transformers/blob/2a5bbdd83b74b61c944093b5d238b70aba196303/x_transformers/x_transformers.py#L794)

**Comment by lucidrains** [comment]
@stoprightthere oh! i forgot i had to change the convention to integrate flash attention for this repository

one more review? :pray:  https://github.com/lucidrains/x-transformers/commit/d731d82b99fbb3491b6a5f0d2012ea2f523411d8

**Comment by lucidrains** [comment]
https://github.com/lucidrains/x-transformers/commit/d2e0700aa0f3648c89885e319636244140d2d602 actually, decided to just redo everything to be consistent

**Comment by stoprightthere** [comment]
looks good!

thanks for the work on this repo and for quick replies!

**Comment by lucidrains** [comment]
@stoprightthere no problem! thanks for identifying this issue and saving a lot of phd students a lot of headache lol
